File: frontend/src/types/loan.ts
export interface BookLoan {
  id: number;
  user_id: number;
  book_id: number;
  book_title: string;
  borrower_name: string;
  borrower_contact: string | null;
  loan_date: string;
  due_date: string | null;
  returned_at: string | null;
  is_returned: boolean;
  is_overdue: boolean;
  notes: string | null;
  created_at: string;
}

export interface CreateLoanPayload {
  book_id: number;
  borrower_name: string;
  borrower_contact?: string | null;
  loan_date?: string | null;
  due_date?: string | null;
  notes?: string | null;
}

File: backend/app/schemas/loan.py
from datetime import date, datetime

from pydantic import BaseModel, Field


class LoanCreate(BaseModel):
    book_id: int
    borrower_name: str = Field(..., min_length=1, max_length=255)
    borrower_contact: str | None = Field(default=None, max_length=255)
    loan_date: date | None = None
    due_date: date | None = None
    notes: str | None = None


class LoanUpdate(BaseModel):
    borrower_name: str | None = Field(default=None, min_length=1, max_length=255)
    borrower_contact: str | None = None
    due_date: date | None = None
    notes: str | None = None


class LoanOut(BaseModel):
    id: int
    user_id: int
    book_id: int
    book_title: str
    borrower_name: str
    borrower_contact: str | None
    loan_date: date
    due_date: date | None
    returned_at: date | None
    is_returned: bool
    is_overdue: bool
    notes: str | None
    created_at: datetime

    model_config = {"from_attributes": True}
